File: Chinpokomon/ChinpokomonJava/src/main/java/poo3/logger/Logger.java
package poo3.logger;

public class Logger {
	private static Logger instance;
    private LoggerLevel level;
    private final String MESSAGE_TEMPLATE = "Logger {level} | {message}";

    private Logger(LoggerLevel level) {
        this.level = level;
    }

    public static Logger getInstance(LoggerLevel level) {
        if (instance == null) {
            instance = new Logger(level);
        }
        
        return instance;
    }
    
    public static Logger getInstance() {
        if (instance == null) {
            instance = new Logger(LoggerLevel.DEBUG);
        }
        
        return instance;
    }
    
    public LoggerLevel getLevel() {
    	return this.level;
    }
    
    public void setLevel(LoggerLevel level) {
    	this.level = level;
    }
    
    private void printMessage(String level, String message) {
    	System.out.println(
    			MESSAGE_TEMPLATE
    				.replace("{level}", level)
    				.replace("{message}", message)	
    	);
    }
    
    private Boolean canPrintMessage(LoggerLevel level) {
    	return this.level.equals(LoggerLevel.DEBUG) || level.ordinal() >= this.level.ordinal();
    }
    
    public void debug(String message) { 
    	if(this.canPrintMessage(LoggerLevel.DEBUG)) {
    		this.printMessage(LoggerLevel.DEBUG.name(), message);
    	}
    }
    
    public void info(String message) { 
    	if(this.canPrintMessage(LoggerLevel.INFO)) {
    		this.printMessage(LoggerLevel.INFO.name(), message);
    	}
    }

    public void warn(String message) { 
    	if(this.canPrintMessage(LoggerLevel.WARN)) {
    		this.printMessage(LoggerLevel.WARN.name(), message);
    	}
    }
    
    public void error(String message) { 
    	if(this.canPrintMessage(LoggerLevel.ERROR)) {
    		this.printMessage(LoggerLevel.ERROR.name(), message);
    	}
    }

}

File: Chinpokomon/ChinpokomonPython/logger/logger.py
from logger.logger_level import LoggerLevel


class SingletonMeta(type):
    _instances = {}

    def __call__(cls, *args, **kwargs):
        if cls not in cls._instances:
            instance = super().__call__(*args, **kwargs)
            cls._instances[cls] = instance
        return cls._instances[cls]


class Logger(metaclass=SingletonMeta):

    def __init__(self, level=LoggerLevel.DEBUG):
        self._level = level
        self._message_template = "Logger {level} | {message}"

    @property
    def level(self):
        return self._level

    @level.setter
    def level(self, level):
        self._level = level

    def print_message(self, level, message):
        print(self._message_template.format(level=level.name, message=message))

    def can_print_message(self, level):
        return self._level is LoggerLevel.DEBUG or level.value >= self._level.value

    def debug(self, message):
        if (self.can_print_message(LoggerLevel.DEBUG)):
            self.print_message(LoggerLevel.DEBUG, message)

    def info(self, message):
        if (self.can_print_message(LoggerLevel.INFO)):
            self.print_message(LoggerLevel.INFO, message)

    def warn(self, message):
        if (self.can_print_message(LoggerLevel.WARN)):
            self.print_message(LoggerLevel.WARN, message)

    def error(self, message):
        if (self.can_print_message(LoggerLevel.ERROR)):
            self.print_message(LoggerLevel.ERROR, message)
